File: sdks/typescript/tests/tier-control.test.ts
/**
 * Unit tests for the SDK tier-control surface (phase13):
 * `deleteByFilter`, `bulkUpdateMetadata`, `copyVectors`, `setVectorExpiry`
 * on `VectorsClient`, and `reencodeCollection`, `setCollectionTtl` on
 * `CollectionsClient`.
 *
 * Verifies the wire shape (URL + body) the SDK emits, plus that the
 * server's response decodes into the expected report types.
 * Pattern mirrors `tests/tier-demotion.test.ts`.
 */

import { describe, it, expect, beforeEach, vi } from 'vitest';
import { VectorsClient } from '../src/client/vectors';
import { CollectionsClient } from '../src/client/collections';
import type { DeleteByFilterReport, BulkUpdateReport, CopyReport, ReencodeJob } from '../src/models/tier-control';

interface MockTransport {
  get: ReturnType<typeof vi.fn>;
  post: ReturnType<typeof vi.fn>;
  put: ReturnType<typeof vi.fn>;
  patch: ReturnType<typeof vi.fn>;
  delete: ReturnType<typeof vi.fn>;
  postFormData: ReturnType<typeof vi.fn>;
}

function createMockTransport(): MockTransport {
  return {
    get: vi.fn(),
    post: vi.fn(),
    put: vi.fn(),
    patch: vi.fn(),
    delete: vi.fn(),
    postFormData: vi.fn(),
  };
}

// ---------------------------------------------------------------------------
// VectorsClient — tier control
// ---------------------------------------------------------------------------

describe('VectorsClient — tier control (phase13)', () => {
  let mockTransport: MockTransport;
  let client: VectorsClient;

  beforeEach(() => {
    mockTransport = createMockTransport();
    client = new VectorsClient({ transport: mockTransport as never });
  });

  describe('deleteByFilter', () => {
    it('posts to /collections/{c}/vectors/delete_by_filter with filter body', async () => {
      const serverReply: DeleteByFilterReport = {
        scanned: 100,
        matched: 3,
        deleted: 2,
        results: [
          { id: 'vec-1', status: 'deleted' },
          { id: 'vec-2', status: 'deleted' },
          { id: 'vec-3', status: 'error', error: 'not found' },
        ],
      };
      mockTransport.post.mockResolvedValue(serverReply);

      const filter = { must: [{ key: 'tier', match: { value: 'cold' } }] };
      const report = await client.deleteByFilter('hot-collection', filter);

      expect(mockTransport.post).toHaveBeenCalledTimes(1);
      expect(mockTransport.post).toHaveBeenCalledWith(
        '/collections/hot-collection/vectors/delete_by_filter',
        { filter },
      );
      expect(report.scanned).toBe(100);
      expect(report.matched).toBe(3);
      expect(report.deleted).toBe(2);
      expect(report.results).toHaveLength(3);
    });

    it('returns the full report from server', async () => {
      const serverReply: DeleteByFilterReport = {
        scanned: 50,
        matched: 50,
        deleted: 50,
        results: [],
      };
      mockTransport.post.mockResolvedValue(serverReply);

      const report = await client.deleteByFilter('my-col', { must: [] });
      expect(report).toEqual(serverReply);
    });
  });

  describe('bulkUpdateMetadata', () => {
    it('posts to /collections/{c}/vectors/bulk_update_metadata with filter and patch', async () => {
      const serverReply: BulkUpdateReport = {
        scanned: 50,
        matched: 5,
        updated: 5,
        results: [{ id: 'vec-1', status: 'updated' }],
      };
      mockTransport.post.mockResolvedValue(serverReply);

      const filter = { must: [{ key: 'source', match: { value: 'legacy' } }] };
      const patch = { migrated: true, source: null };
      const report = await client.bulkUpdateMetadata('my-collection', filter, patch);

      expect(mockTransport.post).toHaveBeenCalledTimes(1);
      expect(mockTransport.post).toHaveBeenCalledWith(
        '/collections/my-collection/vectors/bulk_update_metadata',
        { filter, patch },
      );
      expect(report.updated).toBe(5);
      expect(report.matched).toBe(5);
      expect(report.scanned).toBe(50);
    });

    it('surfaces per-id results without aborting the batch', async () => {
      const serverReply: BulkUpdateReport = {
        scanned: 10,
        matched: 3,
        updated: 2,
        results: [
          { id: 'v1', status: 'updated' },
          { id: 'v2', status: 'updated' },
          { id: 'v3', status: 'error', error: 'locked' },
        ],
      };
      mockTransport.post.mockResolvedValue(serverReply);

      const report = await client.bulkUpdateMetadata('col', {}, {});
      expect(report.updated).toBe(2);
      expect(Array.isArray(report.results)).toBe(true);
      expect(report.results).toHaveLength(3);
    });
  });

  describe('copyVectors', () => {
    it('posts to /collections/{src}/vectors/copy with destination and ids', async () => {
      const serverReply: CopyReport = {
        src: 'hot',
        dst: 'cold',
        requested: 3,
        copied: 2,
        failed: 1,
        results: [
          { id: 'v1', status: 'ok' },
          { id: 'v2', status: 'ok' },
          { id: 'v3', status: 'missing_in_src', error: 'not found' },
        ],
      };
      mockTransport.post.mockResolvedValue(serverReply);

      const report = await client.copyVectors('hot', 'cold', ['v1', 'v2', 'v3']);

      expect(mockTransport.post).toHaveBeenCalledTimes(1);
      expect(mockTransport.post).toHaveBeenCalledWith(
        '/collections/hot/vectors/copy',
        { destination: 'cold', ids: ['v1', 'v2', 'v3'] },
      );
      expect(report.src).toBe('hot');
      expect(report.dst).toBe('cold');
      expect(report.copied).toBe(2);
      expect(report.failed).toBe(1);
    });

    it('returns per-id statuses in results array', async () => {
      const serverReply: CopyReport = {
        src: 'src',
        dst: 'dst',
        requested: 2,
        copied: 1,
        failed: 1,
        results: [
          { id: 'v1', status: 'ok' },
          { id: 'v2', status: 'dst_insert_failed', error: 'dimension mismatch' },
        ],
      };
      mockTransport.post.mockResolvedValue(serverReply);

      const report = await client.copyVectors('src', 'dst', ['v1', 'v2']);
      const statuses = report.results.map((r) => r.status);
      expect(statuses).toEqual(['ok', 'dst_insert_failed']);
    });
  });

  describe('setVectorExpiry', () => {
    it('patches /collections/{c}/vectors/{id}/expiry with expires_at', async () => {
      mockTransport.patch.mockResolvedValue(undefined);

      await client.setVectorExpiry('my-col', 'vec-123', 1893456000000);

      expect(mockTransport.patch).toHaveBeenCalledTimes(1);
      expect(mockTransport.patch).toHaveBeenCalledWith(
        '/collections/my-col/vectors/vec-123/expiry',
        { expires_at: 1893456000000 },
      );
    });

    it('sends null to clear an existing expiry', async () => {
      mockTransport.patch.mockResolvedValue(undefined);

      await client.setVectorExpiry('my-col', 'vec-123', null);

      expect(mockTransport.patch).toHaveBeenCalledWith(
        '/collections/my-col/vectors/vec-123/expiry',
        { expires_at: null },
      );
    });
  });
});

// ---------------------------------------------------------------------------
// CollectionsClient — tier control
// ---------------------------------------------------------------------------

describe('CollectionsClient — tier control (phase13)', () => {
  let mockTransport: MockTransport;
  let client: CollectionsClient;

  beforeEach(() => {
    mockTransport = createMockTransport();
    client = new CollectionsClient({ transport: mockTransport as never });
  });

  describe('reencodeCollection', () => {
    it('posts to /collections/{c}/reencode with target_encoding', async () => {
      const serverReply: ReencodeJob = {
        job_id: 'reencode-my-col-1746000000',
        collection: 'my-col',
        state: 'completed',
        target_encoding: 'sq8',
        progress: 1.0,
      };
      mockTransport.post.mockResolvedValue(serverReply);

      const job = await client.reencodeCollection('my-col', 'sq8');

      expect(mockTransport.post).toHaveBeenCalledTimes(1);
      expect(mockTransport.post).toHaveBeenCalledWith(
        '/collections/my-col/reencode',
        { target_encoding: 'sq8' },
      );
      expect(job.collection).toBe('my-col');
      expect(job.state).toBe('completed');
      expect(job.target_encoding).toBe('sq8');
      expect(job.progress).toBeCloseTo(1.0);
    });

    it('returns the full ReencodeJob from the server', async () => {
      const serverReply: ReencodeJob = {
        job_id: 'reencode-x-99',
        collection: 'x',
        state: 'running',
        target_encoding: 'binary',
        progress: 0.42,
      };
      mockTransport.post.mockResolvedValue(serverReply);

      const job = await client.reencodeCollection('x', 'binary');
      expect(job).toEqual(serverReply);
    });
  });

  describe('setCollectionTtl', () => {
    it('posts to /collections/{c}/ttl with ttl_secs', async () => {
      mockTransport.post.mockResolvedValue(undefined);

      await client.setCollectionTtl('my-col', 3600);

      expect(mockTransport.post).toHaveBeenCalledTimes(1);
      expect(mockTransport.post).toHaveBeenCalledWith(
        '/collections/my-col/ttl',
        { ttl_secs: 3600 },
      );
    });

    it('sends null to clear the collection TTL', async () => {
      mockTransport.post.mockResolvedValue(undefined);

      await client.setCollectionTtl('my-col', null);

      expect(mockTransport.post).toHaveBeenCalledWith(
        '/collections/my-col/ttl',
        { ttl_secs: null },
      );
    });
  });
});

File: sdks/go/tier_control_test.go
package vectorizer

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
)

// TestTierDeleteByFilter verifies that DeleteByFilter POSTs to the correct
// endpoint and deserialises the DeleteByFilterReport response fields.
func TestTierDeleteByFilter(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != "POST" {
			t.Errorf("should use POST method but got %s", r.Method)
		}
		if r.URL.Path != "/collections/c1/vectors/delete_by_filter" {
			t.Errorf("unexpected path: %s", r.URL.Path)
		}

		var body map[string]interface{}
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			t.Fatalf("should receive valid JSON body: %v", err)
		}
		if _, ok := body["filter"]; !ok {
			t.Error("POST body should contain a filter key")
		}

		resp := DeleteByFilterReport{
			Scanned: 10,
			Matched: 4,
			Deleted: 4,
			Results: []VectorOpResult{
				{Status: "ok"},
			},
		}
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(resp)
	}))
	defer server.Close()

	client := NewClient(&Config{BaseURL: server.URL})
	report, err := client.DeleteByFilterRaw("c1", map[string]interface{}{"status": "stale"})
	if err != nil {
		t.Fatalf("DeleteByFilter failed: %v", err)
	}
	if report == nil {
		t.Fatal("should return a non-nil DeleteByFilterReport")
	}
	if report.Scanned != 10 {
		t.Errorf("Scanned should be 10 but got %d", report.Scanned)
	}
	if report.Matched != 4 {
		t.Errorf("Matched should be 4 but got %d", report.Matched)
	}
	if report.Deleted != 4 {
		t.Errorf("Deleted should be 4 but got %d", report.Deleted)
	}
	if len(report.Results) != 1 {
		t.Errorf("Results should have 1 entry but got %d", len(report.Results))
	}
}

// TestTierDeleteByFilterEmptyFails verifies that an empty filter is rejected
// client-side before any HTTP request is issued (raw variant).
func TestTierDeleteByFilterEmptyFails(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		t.Fatal("server should not be called")
	}))
	defer server.Close()

	client := NewClient(&Config{BaseURL: server.URL})

	// nil filter (len == 0) must be rejected client-side.
	_, err := client.DeleteByFilterRaw("c1", nil)
	if err == nil {
		t.Fatal("should return an error for nil filter")
	}
	ve, ok := err.(*VectorizerError)
	if !ok {
		t.Fatalf("error should be *VectorizerError but got %T", err)
	}
	if ve.Type != "validation_error" {
		t.Errorf("error Type should be validation_error but got %q", ve.Type)
	}

	// Empty non-nil map must also be rejected client-side.
	_, err = client.DeleteByFilterRaw("c1", map[string]interface{}{})
	if err == nil {
		t.Fatal("should return an error for empty filter map")
	}
	ve, ok = err.(*VectorizerError)
	if !ok {
		t.Fatalf("error should be *VectorizerError but got %T", err)
	}
	if ve.Type != "validation_error" {
		t.Errorf("error Type should be validation_error but got %q", ve.Type)
	}
}

// TestTierBulkUpdateMetadata verifies that BulkUpdateMetadata POSTs the
// correct body and deserialises the BulkUpdateReport response fields.
func TestTierBulkUpdateMetadata(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != "POST" {
			t.Errorf("should use POST method but got %s", r.Method)
		}
		if r.URL.Path != "/collections/col2/vectors/bulk_update_metadata" {
			t.Errorf("unexpected path: %s", r.URL.Path)
		}

		var body map[string]interface{}
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			t.Fatalf("should receive valid JSON body: %v", err)
		}
		if _, ok := body["filter"]; !ok {
			t.Error("POST body should contain a filter key")
		}
		if _, ok := body["patch"]; !ok {
			t.Error("POST body should contain a patch key")
		}

		resp := BulkUpdateReport{
			Scanned: 5,
			Matched: 3,
			Updated: 3,
			Results: []VectorOpResult{
				{Status: "ok"},
				{Status: "ok"},
				{Status: "ok"},
			},
		}
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(resp)
	}))
	defer server.Close()

	client := NewClient(&Config{BaseURL: server.URL})
	report, err := client.BulkUpdateMetadataRaw(
		"col2",
		map[string]interface{}{"category": "A"},
		map[string]interface{}{"tag": "updated"},
	)
	if err != nil {
		t.Fatalf("BulkUpdateMetadata failed: %v", err)
	}
	if report == nil {
		t.Fatal("should return a non-nil BulkUpdateReport")
	}
	if report.Scanned != 5 {
		t.Errorf("Scanned should be 5 but got %d", report.Scanned)
	}
	if report.Matched != 3 {
		t.Errorf("Matched should be 3 but got %d", report.Matched)
	}
	if report.Updated != 3 {
		t.Errorf("Updated should be 3 but got %d", report.Updated)
	}
}

// TestTierBulkUpdateMetadataEmptyFails verifies that an empty filter is
// rejected client-side before any HTTP request is issued (raw variant).
func TestTierBulkUpdateMetadataEmptyFails(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		t.Fatal("server should not be called")
	}))
	defer server.Close()

	client := NewClient(&Config{BaseURL: server.URL})

	// nil filter must be rejected client-side.
	_, err := client.BulkUpdateMetadataRaw("col2", nil, map[string]interface{}{"tag": "x"})
	if err == nil {
		t.Fatal("should return an error for nil filter")
	}
	ve, ok := err.(*VectorizerError)
	if !ok {
		t.Fatalf("error should be *VectorizerError but got %T", err)
	}
	if ve.Type != "validation_error" {
		t.Errorf("error Type should be validation_error but got %q", ve.Type)
	}

	// Empty non-nil map must also be rejected client-side.
	_, err = client.BulkUpdateMetadataRaw("col2", map[string]interface{}{}, map[string]interface{}{"tag": "x"})
	if err == nil {
		t.Fatal("should return an error for empty filter map")
	}
	ve, ok = err.(*VectorizerError)
	if !ok {
		t.Fatalf("error should be *VectorizerError but got %T", err)
	}
	if ve.Type != "validation_error" {
		t.Errorf("error Type should be validation_error but got %q", ve.Type)
	}
}

// TestTierCopyVectors verifies that CopyVectors POSTs the correct body with
// destination and ids fields.
func TestTierCopyVectors(t *testing.T) {
	var capturedBody map[string]interface{}

	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != "POST" {
			t.Errorf("should use POST method but got %s", r.Method)
		}
		if r.URL.Path != "/collections/src_col/vectors/copy" {
			t.Errorf("unexpected path: %s", r.URL.Path)
		}

		if err := json.NewDecoder(r.Body).Decode(&capturedBody); err != nil {
			t.Fatalf("should receive valid JSON body: %v", err)
		}

		resp := CopyReport{
			Src:       "src_col",
			Dst:       "dst_col",
			Requested: 2,
			Copied:    2,
			Failed:    0,
			Results: []VectorOpResult{
				{Status: "ok"},
				{Status: "ok"},
			},
		}
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(resp)
	}))
	defer server.Close()

	client := NewClient(&Config{BaseURL: server.URL})
	report, err := client.CopyVectors("src_col", "dst_col", []string{"id1", "id2"})
	if err != nil {
		t.Fatalf("CopyVectors failed: %v", err)
	}
	if report == nil {
		t.Fatal("should return a non-nil CopyReport")
	}
	if report.Src != "src_col" {
		t.Errorf("Src should be src_col but got %q", report.Src)
	}
	if report.Dst != "dst_col" {
		t.Errorf("Dst should be dst_col but got %q", report.Dst)
	}
	if report.Requested != 2 {
		t.Errorf("Requested should be 2 but got %d", report.Requested)
	}
	if report.Copied != 2 {
		t.Errorf("Copied should be 2 but got %d", report.Copied)
	}
	if report.Failed != 0 {
		t.Errorf("Failed should be 0 but got %d", report.Failed)
	}

	// Verify POST body shape.
	if capturedBody["destination"] != "dst_col" {
		t.Errorf("POST body destination should be dst_col but got %v", capturedBody["destination"])
	}
	ids, ok := capturedBody["ids"].([]interface{})
	if !ok {
		t.Fatalf("POST body ids should be an array but got %T", capturedBody["ids"])
	}
	if len(ids) != 2 {
		t.Errorf("POST body ids should have 2 elements but got %d", len(ids))
	}
}

// TestTierReencodeCollection verifies that ReencodeCollection POSTs the
// correct body and deserialises the ReencodeJob response.
func TestTierReencodeCollection(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != "POST" {
			t.Errorf("should use POST method but got %s", r.Method)
		}
		if r.URL.Path != "/collections/my_col/reencode" {
			t.Errorf("unexpected path: %s", r.URL.Path)
		}

		var body map[string]interface{}
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			t.Fatalf("should receive valid JSON body: %v", err)
		}
		if body["target_encoding"] != "sq8" {
			t.Errorf("POST body target_encoding should be sq8 but got %v", body["target_encoding"])
		}

		resp := ReencodeJob{
			JobID:          "job-abc",
			Collection:     "my_col",
			State:          "completed",
			TargetEncoding: "sq8",
			Progress:       1.0,
		}
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(resp)
	}))
	defer server.Close()

	client := NewClient(&Config{BaseURL: server.URL})
	job, err := client.ReencodeCollection("my_col", "sq8")
	if err != nil {
		t.Fatalf("ReencodeCollection failed: %v", err)
	}
	if job == nil {
		t.Fatal("should return a non-nil ReencodeJob")
	}
	if job.JobID != "job-abc" {
		t.Errorf("JobID should be job-abc but got %q", job.JobID)
	}
	if job.State != "completed" {
		t.Errorf("State should be completed but got %q", job.State)
	}
	if job.TargetEncoding != "sq8" {
		t.Errorf("TargetEncoding should be sq8 but got %q", job.TargetEncoding)
	}
	if job.Progress != 1.0 {
		t.Errorf("Progress should be 1.0 but got %f", job.Progress)
	}
}

// TestTierSetCollectionTTL verifies that SetCollectionTTL POSTs to the correct
// path and that a nil ttlSecs is serialised as JSON null.
func TestTierSetCollectionTTL(t *testing.T) {
	var capturedBody map[string]interface{}

	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != "POST" {
			t.Errorf("should use POST method but got %s", r.Method)
		}
		if r.URL.Path != "/collections/ttl_col/ttl" {
			t.Errorf("unexpected path: %s", r.URL.Path)
		}

		if err := json.NewDecoder(r.Body).Decode(&capturedBody); err != nil {
			t.Fatalf("should receive valid JSON body: %v", err)
		}

		w.WriteHeader(http.StatusOK)
	}))
	defer server.Close()

	client := NewClient(&Config{BaseURL: server.URL})

	// nil pointer → JSON null (clears TTL).
	err := client.SetCollectionTTL("ttl_col", nil)
	if err != nil {
		t.Fatalf("SetCollectionTTL(nil) failed: %v", err)
	}
	// json.Decoder decodes JSON null as nil interface{}.
	if ttlVal, exists := capturedBody["ttl_secs"]; exists {
		if ttlVal != nil {
			t.Errorf("ttl_secs should be null when pointer is nil but got %v", ttlVal)
		}
	} else {
		t.Error("POST body should contain ttl_secs key")
	}

	// Non-nil pointer → concrete integer value.
	var ttlSecs int64 = 3600
	err = client.SetCollectionTTL("ttl_col", &ttlSecs)
	if err != nil {
		t.Fatalf("SetCollectionTTL(3600) failed: %v", err)
	}
	// JSON numbers decode as float64 in map[string]interface{}.
	if capturedBody["ttl_secs"] != float64(3600) {
		t.Errorf("ttl_secs should be 3600 but got %v", capturedBody["ttl_secs"])
	}
}

// TestTierSetVectorExpiry verifies that SetVectorExpiry uses PATCH, targets
// the correct path, and serialises a nil expiresAt as JSON null.
func TestTierSetVectorExpiry(t *testing.T) {
	var capturedBody map[string]interface{}
	var capturedMethod string

	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		capturedMethod = r.Method
		if r.URL.Path != "/collections/exp_col/vectors/vec1/expiry" {
			t.Errorf("unexpected path: %s", r.URL.Path)
		}

		if err := json.NewDecoder(r.Body).Decode(&capturedBody); err != nil {
			t.Fatalf("should receive valid JSON body: %v", err)
		}

		w.WriteHeader(http.StatusOK)
	}))
	defer server.Close()

	client := NewClient(&Config{BaseURL: server.URL})

	// nil pointer → JSON null (clears expiry).
	err := client.SetVectorExpiry("exp_col", "vec1", nil)
	if err != nil {
		t.Fatalf("SetVectorExpiry(nil) failed: %v", err)
	}
	if capturedMethod != "PATCH" {
		t.Errorf("should use PATCH method but got %s", capturedMethod)
	}
	if expiresVal, exists := capturedBody["expires_at"]; exists {
		if expiresVal != nil {
			t.Errorf("expires_at should be null when pointer is nil but got %v", expiresVal)
		}
	} else {
		t.Error("POST body should contain expires_at key")
	}

	// Non-nil pointer → concrete integer value.
	var expiresAt int64 = 9999999999
	err = client.SetVectorExpiry("exp_col", "vec1", &expiresAt)
	if err != nil {
		t.Fatalf("SetVectorExpiry(9999999999) failed: %v", err)
	}
	if capturedMethod != "PATCH" {
		t.Errorf("should use PATCH method but got %s", capturedMethod)
	}
	if capturedBody["expires_at"] != float64(9999999999) {
		t.Errorf("expires_at should be 9999999999 but got %v", capturedBody["expires_at"])
	}
}

// TestTierDeleteByFilterTypedHappy verifies that DeleteByFilter (typed) POSTs
// the correct JSON body shape and decodes the response correctly.
func TestTierDeleteByFilterTypedHappy(t *testing.T) {
	var capturedFilter map[string]interface{}

	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != "POST" {
			t.Errorf("should use POST method but got %s", r.Method)
		}
		if r.URL.Path != "/collections/typed_col/vectors/delete_by_filter" {
			t.Errorf("unexpected path: %s", r.URL.Path)
		}

		var body map[string]interface{}
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			t.Fatalf("should receive valid JSON body: %v", err)
		}

		filterVal, ok := body["filter"]
		if !ok {
			t.Fatal("POST body should contain a filter key")
		}
		capturedFilter, ok = filterVal.(map[string]interface{})
		if !ok {
			t.Fatalf("filter should be an object but got %T", filterVal)
		}

		resp := DeleteByFilterReport{
			Scanned: 8,
			Matched: 2,
			Deleted: 2,
			Results: []VectorOpResult{{Status: "ok"}, {Status: "ok"}},
		}
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(resp)
	}))
	defer server.Close()

	client := NewClient(&Config{BaseURL: server.URL})
	f := MustFilter(FilterEq("status", "stale"))
	report, err := client.DeleteByFilter("typed_col", &f)
	if err != nil {
		t.Fatalf("DeleteByFilter typed failed: %v", err)
	}
	if report == nil {
		t.Fatal("should return a non-nil DeleteByFilterReport")
	}
	if report.Deleted != 2 {
		t.Errorf("Deleted should be 2 but got %d", report.Deleted)
	}

	// Assert the wire shape: filter.must[0].key == "status",
	// filter.must[0].match.value == "stale".
	mustArr, ok := capturedFilter["must"].([]interface{})
	if !ok || len(mustArr) == 0 {
		t.Fatalf("filter.must should be a non-empty array but got %v", capturedFilter["must"])
	}
	cond, ok := mustArr[0].(map[string]interface{})
	if !ok {
		t.Fatalf("must[0] should be an object but got %T", mustArr[0])
	}
	if cond["key"] != "status" {
		t.Errorf("must[0].key should be 'status' but got %v", cond["key"])
	}
	matchObj, ok := cond["match"].(map[string]interface{})
	if !ok {
		t.Fatalf("must[0].match should be an object but got %T", cond["match"])
	}
	if matchObj["value"] != "stale" {
		t.Errorf("must[0].match.value should be 'stale' but got %v", matchObj["value"])
	}
}

// TestTierDeleteByFilterTypedEmptyRejectedClientSide verifies that a nil
// filter and an empty *QdrantFilter are both rejected client-side without
// touching the server.
func TestTierDeleteByFilterTypedEmptyRejectedClientSide(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		t.Fatal("server should not be called for empty typed filter")
	}))
	defer server.Close()

	client := NewClient(&Config{BaseURL: server.URL})

	// nil pointer must be rejected client-side.
	_, err := client.DeleteByFilter("col", nil)
	if err == nil {
		t.Fatal("should return an error for nil *QdrantFilter")
	}
	ve, ok := err.(*VectorizerError)
	if !ok {
		t.Fatalf("error should be *VectorizerError but got %T", err)
	}
	if ve.Type != "validation_error" {
		t.Errorf("error Type should be validation_error but got %q", ve.Type)
	}

	// Empty (no conditions) filter must also be rejected client-side.
	empty := &QdrantFilter{}
	_, err = client.DeleteByFilter("col", empty)
	if err == nil {
		t.Fatal("should return an error for empty *QdrantFilter")
	}
	ve, ok = err.(*VectorizerError)
	if !ok {
		t.Fatalf("error should be *VectorizerError but got %T", err)
	}
	if ve.Type != "validation_error" {
		t.Errorf("error Type should be validation_error but got %q", ve.Type)
	}
}
